In [44]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error,mean_squared_error,r2_score
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.ensemble import RandomForestRegressor,GradientBoostingRegressor
from sklearn.impute import SimpleImputer
from sklearn.svm import SVR
from sklearn.compose import ColumnTransformer

In [45]:
data = pd.DataFrame({
    "EngineSize": [1.0, 1.2, 1.4, 1.6, 1.8, np.nan, 2.0, 2.2, 2.4, 2.6],
    "HorsePower": [68, 75, 90, 100, 110, 115, np.nan, 130, 150, 160],
    "FuelType": ["Petrol", "Diesel", "Petrol", "Diesel", "Petrol",
                 "Petrol", np.nan, "Diesel", "Petrol", "Diesel"],
    "FuelEfficiency": [20, 22, 19, 25, 18, 21, 23, 26, 17, 24]
})
data

,EngineSize,HorsePower,FuelType,FuelEfficiency
0,1.0,68.0,Petrol,20
1,1.2,75.0,Diesel,22
2,1.4,90.0,Petrol,19
3,1.6,100.0,Diesel,25
4,1.8,110.0,Petrol,18
5,NaN,115.0,Petrol,21
6,2.0,NaN,NaN,23
7,2.2,130.0,Diesel,26
8,2.4,150.0,Petrol,17
9,2.6,160.0,Diesel,24


In [46]:
X = data[["EngineSize", "HorsePower", "FuelType"]]
y = data["FuelEfficiency"]
data

,EngineSize,HorsePower,FuelType,FuelEfficiency
0,1.0,68.0,Petrol,20
1,1.2,75.0,Diesel,22
2,1.4,90.0,Petrol,19
3,1.6,100.0,Diesel,25
4,1.8,110.0,Petrol,18
5,NaN,115.0,Petrol,21
6,2.0,NaN,NaN,23
7,2.2,130.0,Diesel,26
8,2.4,150.0,Petrol,17
9,2.6,160.0,Diesel,24


In [47]:
num_cols=["EngineSize","HorsePower"]
cat_cols=["FuelType"]

In [48]:
numeric=Pipeline([
    ("imputer",SimpleImputer(strategy="median")),
    ("scaler",StandardScaler())
])
categorical=Pipeline([
    ("imputer",SimpleImputer(strategy="most_frequent")),
    ("onehot",OneHotEncoder(handle_unknown="ignore"))
])

In [49]:
preprocess=ColumnTransformer([("num",numeric,num_cols),("cat",categorical,cat_cols)])

In [53]:
model=Pipeline([
    ("preprocess",preprocess),
    ("regressor",LinearRegression())
    #("regressor",GradientBoostingRegressor(n_estimators=150,random_state=42,max_depth=3,learning_rate=0.1))
    #("regressor",SVR(kernel='rbf',C=100,gamma=0.1))
    #("model", RandomForestRegressor(n_estimators=100,random_state=42))
])

In [54]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)
model.fit(X_train,y_train)

Pipeline(steps=[('preprocess',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='median')),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['EngineSize', 'HorsePower']),
                                                 ('cat',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  ('onehot',
                                                                   OneHotEncoder(handle_unknown='ignore'))]),
                                                  ['FuelType'])])),
                ('regressor', LinearRegression())])

In [55]:
y_pred=model.predict(X_test)
print(y_pred)

[18.88550383 25.54413659]


In [56]:
mae=mean_absolute_error(y_test,y_pred)

rmse=np.sqrt(mean_squared_error(y_test,y_pred))
r2=r2_score(y_test,y_pred)
print("Mean Absolute Error:",round(mae,2))

print("Root Mean Squared Error:",round(rmse,2))
print("R2 Score",round(r2,2))

Mean Absolute Error: 2.71
Root Mean Squared Error: 2.84
R2 Score -0.29


# this is evaluationof Linear Regression
Mean Absolute Error: 2.71
Root Mean Squared Error: 2.84
R2 Score -0.29

In [57]:
mae=mean_absolute_error(y_test,y_pred)
mse=mean_squared_error(y_test,y_pred)
rmse=np.sqrt(mean_squared_error(y_test,y_pred))
r2=r2_score(y_test,y_pred)
print("Mean Absolute Error:",round(mae,2))
print("MSE:",mse)
print("Root Mean Squared Error:",round(rmse,2))
print("R2 Score",round(r2,2))

Mean Absolute Error: 2.71
MSE: 8.05801444891766
Root Mean Squared Error: 2.84
R2 Score -0.29


Mean Absolute Error: 3.16
MSE: 17.21705000000001
Root Mean Squared Error: 4.15
R2 Score -1.75

In [58]:
mae=mean_absolute_error(y_test,y_pred)
mse=mean_squared_error(y_test,y_pred)
rmse=np.sqrt(mean_squared_error(y_test,y_pred))
r2=r2_score(y_test,y_pred)
print("Mean Absolute Error:",round(mae,2))
print("MSE:",mse)
print("Root Mean Squared Error:",round(rmse,2))
print("R2 Score",round(r2,2))

Mean Absolute Error: 2.71
MSE: 8.05801444891766
Root Mean Squared Error: 2.84
R2 Score -0.29
